nixos/modules/tasks/auto-upgrade.nix

```diff
@@ -158,113 +180,185 @@ in {
+
+                lower="${cfg.rebootWindow.lower}"
+                upper="${cfg.rebootWindow.upper}"
+
+                if [[ "''${lower}" < "''${upper}" ]]; then
+                  if [[ "''${current_time}" > "''${lower}" ]] && \
+                     [[ "''${current_time}" < "''${upper}" ]]; then
+                    do_reboot="true"
+                  else
+                    do_reboot="false"
+                  fi
+                else
+                  # lower > upper, so we are crossing midnight (e.g. lower=23h, upper=6h)
+                  # we want to reboot if cur > 23h or cur < 6h
+                  if [[ "''${current_time}" < "''${upper}" ]] || \
+                     [[ "''${current_time}" > "''${lower}" ]]; then
+                    do_reboot="true"
+                  else
+                    do_reboot="false"
+                  fi
+                fi
+              ''}
+
+              if [ "''${booted}" = "''${built}" ]; then
+                ${nixos-rebuild} ${cfg.operation} ${toString cfg.flags}
+              ${lib.optionalString (cfg.rebootWindow != null) ''
+                elif [ "''${do_reboot}" != true ]; then
+                  echo "Outside of configured reboot window, skipping."
+              ''}
+              else
+                ${shutdown} -r +1
+              fi
+            ''
+          else
+            ''
+              ${nixos-rebuild} ${cfg.operation} ${toString (cfg.flags ++ upgradeFlag)}
+            '';
 
-      after = [ "network-online.target" ];
-      wants = [ "network-online.target" ];
-    };
+        startAt = cfg.dates;
 
-    systemd.timers.nixos-upgrade = {
-      timerConfig = {
-        RandomizedDelaySec = cfg.randomizedDelaySec;
-        FixedRandomDelay = cfg.fixedRandomDelay;
-        Persistent = cfg.persistent;
+        after = [ "network-online.target" ];
+        wants = [ "network-online.target" ];
       };
-    };
-  };
 
+      systemd.timers.nixos-upgrade = {
+        timerConfig = {
+          RandomizedDelaySec = cfg.randomizedDelaySec;
+          FixedRandomDelay = cfg.fixedRandomDelay;
+          Persistent = cfg.persistent;
+        };
+      };
+    })
+
+    (lib.mkIf cfg.onShutdown {
+      systemd.services.nixos-upgrade-shutdown = {
+        description = "NixOS Upgrade on Shutdown";
+
+        unitConfig = {
+          DefaultDependencies = false;
+          Before = [
+            "shutdown.target"
+            "reboot.target"
+          ];
+          Conflicts = [
+            "shutdown.target"
+            "reboot.target"
+          ];
+        };
+
+        serviceConfig = {
+          Type = "oneshot";
+          RemainAfterExit = true;
+          User = "root";
+        };
+
+        environment =
+          config.nix.envVars
+          // {
+            inherit (config.environment.sessionVariables) NIX_PATH;
+            HOME = "/root";
+          }
+          // config.networking.proxy.envVars;
+
+        path = with pkgs; [
```

Comment:
I'd prefer avoiding `with`, especially in this situation, because some of the entries in the list are not from `pkgs`.